{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STAR-HTSeq-DEseq2 pipeline for RNAseq analysis\n",
    "\n",
    "### dataset: 5 paired-end RNAseq \n",
    "             2 control and 3 experiment samples\n",
    "             sequence depth: ~50M reads/sample\n",
    "\n",
    "### STAR alignment: 15-20 minutes/full dataset (8 threads)\n",
    "### HTseq: ~2-3 hours/full dataset (1 thread)   ~5 minutes/test files of 400K SAM reads       \n",
    "### DEseq2: "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STAR\n",
    "\n",
    "## genome index\n",
    "__1. generate your own genome index(recommended)\n",
    "\n",
    "   a. download the genome fasta files\n",
    "   b. run STAR in genome index mode\n",
    "   e.g. \n",
    "   STAR --runThreadN 4 --runMode genomeGenerate --genomeDir Genome_data/star \\\n",
    "   --genomeFastaFiles Genome_data/Mus_musculus_UCSC_mm10.tar.gz\n",
    "   \n",
    "   \n",
    "__2. download from Star genome (human and mouse)\n",
    "\n",
    "   link:http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STARgenomes/\n",
    "   \n",
    "   star genome index contains: {chrLength.txt  chrName.txt      chrStart.txt    sjdbInfo.txt    Genome  \n",
    "    genomeParameters.txt   SA    SAindex}\n",
    "\n",
    "## STAR command and options\n",
    "   \n",
    "   STAR --option1-name option1-value(s) --option2-name option2-value(s)....\n",
    "   \n",
    "   _1. required\n",
    "    --runThreadN NumberOfThreads\n",
    "    --genomeDir /path/to/genomeDir\n",
    "    --readFilesIn /path/to/read1 [/path/to/read2]\n",
    "    --outFileNamePrefix /path/to/output/dir/prefix\n",
    "    \n",
    "   _2. Optional\n",
    "    --readFilesCommand UncompressionCommand e.g.zcat - to uncompress .gz files\n",
    "    --outSAMmapqUnique Integer0to255 e.g. mapping quality MAPQ=255 for the unique mappers\n",
    "    --outSAMtype SAM/BAM Unsorted/SortedByCoordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# extract path information for raw fastq and output from config file and store them in lists\n",
    "file_read = open(\"STAR_align.config\", \"r\")\n",
    "fastq_list = []\n",
    "output_list = []\n",
    "\n",
    "for line in file_read:\n",
    "    if line.startswith(\"RAWfastq>\"):\n",
    "        line=line.strip()\n",
    "        line=line[9:]\n",
    "        fastq_list.append(line)\n",
    "    if line.startswith(\"output>\"):\n",
    "        line=line.strip()\n",
    "        line=line[7:]\n",
    "        output_list.append(line)\n",
    "        \n",
    "print(fastq_list)\n",
    "print(output_list)\n",
    "file_read.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## how to pass variables from python to bash: use $"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# use for loop to go through all the files in STAR and store the output files in respective directories.\n",
    "dataset_count =0\n",
    "for index in range (0,len(fastq_list),2):\n",
    "    read_1 = fastq_list[index]\n",
    "    read_2 = fastq_list[index+1]\n",
    "    output_path = output_list[dataset_count]\n",
    "    dataset_count += 1\n",
    "    # print(read_1, read_2, output_path)\n",
    "    ! time STAR --runThreadN 8 --genomeDir /data/project/refgenome_STAR  --readFilesIn $read_1 $read_2 \\\n",
    "    --outSAMmapqUnique 255 -- outSAMtype BAM SortedByCoordinate \\\n",
    "    --outFileNamePrefix $output_path\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HTseq\n",
    "\n",
    "   \n",
    "   ## 1. gene annotation - GTF or GFF file for HTseq count funciton\n",
    "         e.g. human Gencode V25 - comprehansive gene annotations\n",
    "         link: http://www.gencodegenes.org/releases/25.html\n",
    "   \n",
    "   ## 2. sorted BAM files from STAR aligner\n",
    "         note: STAR sort BAM based on mapped read coordinates\n",
    "         \n",
    "   ## 3. htseq-count: \n",
    "         more information: http://htseq.readthedocs.io/en/release_0.9.1/count.html?highlight=htseq-count\n",
    "   \n",
    "         Basic command line: htseq-count [options] <alignment_files> <gff_file> > htseq_output_countfile\n",
    "         options:\n",
    "         -f <format>, --format=<format> sam/bam\n",
    "         -r <order>, --order=<order> name/pros (aligned read ID/positions)\n",
    "                     for paired end reads, alignment files have to be sorted.\n",
    "         \n",
    "         -i <id attribute>, --idattr=<id attribute> e.g. gene_id, gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# extract path information for raw fastq and output from config file and store them in lists\n",
    "\n",
    "file_read = open(\"HTseq.config\", \"r\")\n",
    "input_dir = []\n",
    "output_dir = \"\"\n",
    "output_name = []\n",
    "output_group = []\n",
    "\n",
    "for line in file_read:\n",
    "    if line.startswith(\"inputBAM>\"):  #parse path for BAM input file\n",
    "        line = line[9:]\n",
    "        line = line.strip()\n",
    "        input_dir.append(line)\n",
    "    if line.startswith(\"outputpath>\"): #parse path for output HTseq files\n",
    "        line=line.strip()\n",
    "        output_dir = line[11:]\n",
    "    if line.startswith(\"groupinfo>\"):  #parse dataset info (dataID and group/condition)\n",
    "        line = line[10:]\n",
    "        output = line.strip().split(\"_\")\n",
    "        output_name.append(output[0])\n",
    "        output_group.append(output[1])\n",
    "#print(input_dir)\n",
    "#print(output_dir)\n",
    "#print(output_name)\n",
    "#print(output_group)\n",
    "file_read.close()\n",
    "\n",
    "\n",
    "file_write = open (\"/data/project/HTseq/deseq2_inputsummary.txt\", \"w\") # initiate a meta file table for downstream DESeq\n",
    "file_write.write(\"sampleID\\t\"+\"filenames\\t\"+\"conditions\"+\"\\n\")\n",
    "\n",
    "for i in range(0,len(input_dir)):\n",
    "    input_path = input_dir[i]\n",
    "    output_filename = output_name[i]+\"_\"+output_group[i]+\"_output.txt\"\n",
    "    output_path_name = str(output_dir)+output_filename\n",
    "    #print(input_path)\n",
    "    #print(output_filename)\n",
    "    #print(output_path_name)\n",
    "    file_write.write(output_name[i]+\"\\t\"+output_filename+\"\\t\"+output_group[i]+\"\\n\")\n",
    "    \n",
    "    ! time htseq-count -f bam -r pos -i gene_name \\\n",
    "    $input_path GencodeV25_GTF/CHR_comprehansive_V25.gtf > $output_path_name\n",
    "\n",
    "file_write.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! Rscript /data/project/HTseq/DEseq2_test2.r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DEseq2_test2.r\n",
    "\n",
    "\n",
    "meta <- read.table(\"/data/project/HTseq/htseq_test2/deseq2_inputsummary.txt\", header =TRUE)\n",
    "\n",
    "#this summary table was generated from HTSeq run and contain sampleID, HTSeq count output file names and experimental conditions of your datasets.\n",
    "#Note: the directory of these files is assigned specifically as DEseqDataSetFromHTSeqCount variable. Don't put path info here in the meta table.\n",
    "\n",
    "\n",
    "#assign attributes of sampleTable based on meta table\n",
    "\n",
    "samplenames <- as.character(meta$sampleID)\n",
    "filenames <- as.character(meta$filenames)\n",
    "samplecondition <- as.character(meta$condition)\n",
    "\n",
    "sampleTable <- data.frame(sampleName = samplenames, fileName = filenames, condition = samplecondition)\n",
    "print(sampleTable)\n",
    "\n",
    "\n",
    "library(DESeq2) # Call DESeq2 in R and input HTseq count\n",
    "\n",
    "directory <- \"/data/project/HTseq/htseq_test2/\" # assign directory of HTseq count output files\n",
    "\n",
    "\n",
    "# assign the sample table to DESeq function for HTseq count input\n",
    "\n",
    "ddsHTSeq <- DESeqDataSetFromHTSeqCount (sampleTable = sampleTable, directory = directory, design = ~ condition)\n",
    "print(ddsHTSeq)\n",
    "\n",
    "dds <- DESeq(ddsHTSeq)\n",
    "res <- results(dds)  \n",
    "res05 <- results(dds,alpha=0.05) # highlight DE genes with FDR <0.05\n",
    "summary(res05) \n",
    "sum(res05$padj <0.05, na.rm=TRUE) # how many DE genes meet the threshhold\n",
    "\n",
    "plotMA(res05, ylim=c(-10,10)) # MAplot\n",
    "jpeg(\"DEseq_res05_MAplot2.jpg\")\n",
    "plotMA(res05, ylim=c(-10,10))\n",
    "dev.off()\n",
    "\n",
    "res05Ordered <- res05[order(res05$pvalue),] \n",
    "write.csv(as.data.frame(res05Ordered), file=\"parental_brain_FDR05_results.csv\") # export results into csv fles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  sampleName                 fileName condition\n",
    "1 SRR3649333 htseq_test333_output.txt  parental\n",
    "2 SRR3649334 htseq_test334_output.txt  parental\n",
    "3 SRR3649335 htseq_test335_output.txt     brain\n",
    "4 SRR3649336 htseq_test336_output.txt     brain\n",
    "5 SRR3649337 htseq_test337_output.txt     brain\n",
    "Loading required package: S4Vectors\n",
    "Loading required package: methods\n",
    "Loading required package: stats4\n",
    "Loading required package: BiocGenerics\n",
    "Loading required package: parallel\n",
    "\n",
    "Attaching package: ‘BiocGenerics’\n",
    "\n",
    "The following objects are masked from ‘package:parallel’:\n",
    "\n",
    "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
    "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
    "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
    "\n",
    "The following objects are masked from ‘package:stats’:\n",
    "\n",
    "    IQR, mad, xtabs\n",
    "\n",
    "The following objects are masked from ‘package:base’:\n",
    "\n",
    "    anyDuplicated, append, as.data.frame, as.vector, cbind, colnames,\n",
    "    do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl,\n",
    "    intersect, is.unsorted, lapply, lengths, Map, mapply, match, mget,\n",
    "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
    "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
    "    union, unique, unlist, unsplit\n",
    "\n",
    "Loading required package: IRanges\n",
    "Loading required package: GenomicRanges\n",
    "Loading required package: GenomeInfoDb\n",
    "Loading required package: SummarizedExperiment\n",
    "Loading required package: Biobase\n",
    "Welcome to Bioconductor\n",
    "\n",
    "    Vignettes contain introductory material; view with\n",
    "    'browseVignettes()'. To cite Bioconductor, see\n",
    "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
    "\n",
    "Loading required package: Rcpp\n",
    "Loading required package: RcppArmadillo\n",
    "class: DESeqDataSet \n",
    "dim: 56269 5 \n",
    "metadata(0):\n",
    "assays(1): counts\n",
    "rownames(56269): 5S_rRNA 5_8S_rRNA ... uc_338 yR211F11.2\n",
    "rowRanges metadata column names(0):\n",
    "colnames(5): SRR3649333 SRR3649334 SRR3649335 SRR3649336 SRR3649337\n",
    "colData names(1): condition\n",
    "estimating size factors\n",
    "estimating dispersions\n",
    "gene-wise dispersion estimates\n",
    "mean-dispersion relationship\n",
    "final dispersion estimates\n",
    "fitting model and testing\n",
    "\n",
    "out of 95 with nonzero total read count\n",
    "adjusted p-value < 0.05\n",
    "LFC > 0 (up)     : 6, 6.3% \n",
    "LFC < 0 (down)   : 0, 0% \n",
    "outliers [1]     : 0, 0% \n",
    "low counts [2]   : 0, 0% \n",
    "(mean count < 0)\n",
    "[1] see 'cooksCutoff' argument of ?results\n",
    "[2] see 'independentFiltering' argument of ?results\n",
    "\n",
    "[1] 6\n",
    "pdf \n",
    "  2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
